docs/question/server.md

```diff
@@ -1,17 +1,104 @@
+- `data["submitted_answers"][NAME]` is the answer parsed by elements (e.g. strings converted to numbers).
+- `data["format_errors"][NAME]` is the format error message for a named answer.
+
+If there are format errors, then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.
+
+```python title="server.py"
+def parse(data):
+    # Reject negative numbers for "y" if we don't already have a format error
+    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+        data["format_errors"]["y"] = "Negative numbers are not allowed"
+```
+
+### Step 5: `grade`
+
+Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:
+
+- Setting the score and feedback for each named answer in `data["partial_scores"]`.
+- Setting the total score for the question in `data["score"]`.
+- Setting the overall feedback for the question in `data["feedback"]`.
+
+This function only runs if `parse()` did not produce format errors, so we can assume all data is valid. All elements will have already graded their answers (if any) before this point.
+
+- `data["partial_scores"][NAME]["score"]` is the individual scores for the named answers (0 to 1).
+- `data["score"]` is the total score for the question (0 to 1).
+- `data["feedback"][NAME]` is the feedback for each named answer.
+
+It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.
+
+If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer.
+
+You can set `data["format_errors"][NAME]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments. However, you are encouraged to do as many checks for invalid data as possible in `parse` instead of `grade`; the `parse` function is called when the student hits "Save only", in manually graded questions, and in assessments without real-time grading.
+
+```python title="server.py"
+import math
+
+def grade(data):
+    # Give half points for incorrect answers larger than "x", only if not already correct.
+    # Use math.isclose to avoid possible floating point errors.
+    y_is_correct = math.isclose(data["partial_scores"]["y"]["score"], 1.0)
+    if not y_is_correct and data["submitted_answers"]["y"] > data["params"]["x"]:
+        data["partial_scores"]["y"]["score"] = 0.5
+        data["score"] = set_weighted_score_data(data)
```

Comment:
`set_weighted_score_data()` updates `data["score"]` directly, and does not return anything. You also need to import it.

Reply:
Fixed in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9.